### ISSUE TITLE
Bug: bouncer.go:176:6: panic

### ISSUE THREAD

**Issue by mpsarakis** [maxlerebourg/crowdsec-bouncer-traefik-plugin#60]
Hello,

I am a new user of the traefik bouncer plugin in a (stable) docker environment and I have unexplained errors in the logs:
time="2022-12-06T18:33:12+01:00" level=error msg="plugins-storage/sources/gop-3552147530/src/github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/bouncer.go:176:6: panic" plugin=plugin-bouncer module=github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin

I have activated the DEBUG log level but nothing appears in the logs, error surrounding lines are these ones but they also appear outside of the error context so I don't think they are directly related:
DEBUG: CrowdsecBouncerTraefikPlugin: 2022/12/06 18:33:12 ServeHTTP ip:82.xxx.xxx.xxx isTrusted:true

Apart from that, everything seems to be working (as far as my recent experience of the product goes) and I don't know what this error causes if anything.

I have found no way to reproduce the error.

Traefik version 2.9.5
Plugin version 1.1.5

Thanks.

Plugin configuration:
    middlewares-traefik-bouncer:
      plugin:
        bouncer:
          enabled: true
          logLevel: DEBUG
          crowdsecMode: live
          crowdsecLapiScheme: http
          crowdsecLapiHost: crowdsec:8080
          crowdsecLapiKey: xxxx
          clientTrustedIPs: 
            - 192.168.0.0/24
            - 82.xxx.xxx.xxx
          forwardedHeadersTrustedIPs: 
            # Cloudflare IP Ranges
            # https://www.cloudflare.com/ips/
            - 173.245.48.0/20
            - 103.21.244.0/22
            - 103.22.200.0/22
            - 103.31.4.0/22
            - 141.101.64.0/18
            - 108.162.192.0/18
            - 190.93.240.0/20
            - 188.114.96.0/20
            - 197.234.240.0/22
            - 198.41.128.0/17
            - 162.158.0.0/15
            - 104.16.0.0/13
            - 104.24.0.0/14
            - 172.64.0.0/13
            - 131.0.72.0/22

**Comment by mathieuHa** [comment]
Hello @mpsarakis,

First thanks for using this plugin.

We'll try to figure out what's going on.

Thank you for the information provided, that will help.

I see that you are using version **1.1.5** which is a **pre-release** version. 
We with @maxlerebourg heavily test release version which we consider to be production ready.

Could you try to use the plugin with version **1.1.4** and let us know if that works without any issue ?

We will investigate in the next days to find out what could have cause this issue.

Could you post the starting logs (in debug mode) when Traefik restart ?
This might give us a hint on what happened.

We'll keep you updated,

Mathieu

**Comment by mathieuHa** [comment]
Hey Marios,

With the information provided I've been able to reproduce what you are experiencing.
Dozzle indeed could trigger the error.

```
traefik     | DEBUG: CrowdsecBouncerTraefikPlugin: 2022/12/08 20:16:28 ServeHTTP ip:10.0.XX.XX isTrusted:false
traefik     | DEBUG: CrowdsecBouncerTraefikPlugin: 2022/12/08 20:16:28 ServeHTTP ip:10.0.XX.XX cache:hit isBanned:false
traefik     | time="2022-12-08T20:16:29Z" level=error msg="plugins-storage/sources/gop-2455797364/src/github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/bouncer.go:176:6: panic" plugin=plugin-bouncer module=github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin
traefik     | DEBUG: CrowdsecBouncerTraefikPlugin: 2022/12/08 20:16:29 ServeHTTP ip:10.0.XX.XX isTrusted:false
```

Here is the following docker-compose that can help reproduce:

```
version: "3.8"

services:
  traefik:
    image: "traefik:v2.9.5"
    container_name: "traefik"
    restart: unless-stopped
    command:
      - "--accesslog"
      - "--accesslog.filepath=/var/log/traefik/access.log"
      - "--api.insecure=true"
      - "--providers.docker=true"
      - "--providers.docker.exposedbydefault=false"
      - "--entrypoints.web.address=:80"

      - "--experimental.plugins.bouncer.modulename=github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin"
      - "--experimental.plugins.bouncer.version=v1.1.5"
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
      - "logs:/var/log/traefik"
    ports:
      - 8000:80
      - 8080:8080
    depends_on:
      - 'crowdsec'

  crowdsec:
    image: crowdsecurity/crowdsec:v1.4.1
    container_name: "crowdsec"
    restart: unless-stopped
    environment:
      COLLECTIONS: crowdsecurity/traefik
      CUSTOM_HOSTNAME: crowdsec
      # We need to register one api key per service we will use
      BOUNCER_KEY_TRAEFIK_1: FIXME-LAPI-KEY-1
    volumes:
      - ./acquis.yaml:/etc/crowdsec/acquis.yaml:ro
      - logs:/var/log/traefik:ro
      - crowdsec-db:/var/lib/crowdsec/data/
      - crowdsec-config:/etc/crowdsec/
    labels:
      - "traefik.enable=false"
  
  dozzle:
    container_name: dozzle
    image: amir20/dozzle:latest
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.router-doz.rule=Host(`localhost`)"
      - "traefik.http.routers.router-doz.entrypoints=web"
      - "traefik.http.routers.router-doz.middlewares=crowdsec-doz@docker" 
      - "traefik.http.services.service-doz.loadbalancer.server.port=8080"
      - "traefik.http.middlewares.crowdsec-doz.plugin.bouncer.enabled=true"
      - "traefik.http.middlewares.crowdsec-doz.plugin.bouncer.crowdseclapikey=FIXME-LAPI-KEY-1"
      - "traefik.http.middlewares.crowdsec-doz.plugin.bouncer.loglevel=DEBUG"
      - "traefik.http.middlewares.crowdsec-doz.plugin.bouncer.forwardedheaderstrustedips=173.245.48.0/20,103.21.244.0/22,103.22.200.0/22"


volumes:
  logs:
  crowdsec-db:
  crowdsec-config:
```

I'll investigate even deeper in the following days.

Even if it said panic in the log, did you encounter any bad behavior ?
It's seems to be "working" although it says otherwise in the log.

Best,
Mathieu

**Comment by mathieuHa** [comment]
It's actually pretty weird,
When I set the plugin mode to disable, I still get the error showed.

``` - "traefik.http.middlewares.crowdsec-doz.plugin.bouncer.enabled=false" ```
In disabled mode, the plugin only do one thing, transmit the request to the next handler / middleware.

```
if !bouncer.enabled {
  bouncer.next.ServeHTTP(rw, req)
  return
}
```

I'm starting to believe It could be an issue in Traefik itself.

**Comment by mathieuHa** [comment]
With Traefik in Debug mode, I can observe one more interresting log:

```
time="2022-12-08T20:48:05Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/api/logs/stream\",\"RawPath\":\"\",\"OmitHost\":false,\"ForceQuery\":false,\"RawQuery\":\"id=dfdf19c9d502\\u0026lastEventId=\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"text/event-stream\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Accept-Language\":[\"fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Dnt\":[\"1\"],\"Pragma\":[\"no-cache\"],\"Referer\":[\"http://10.0.10.XX:8000/container/dfdf19c9d502\"],\"User-Agent\":[\"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36\"],\"X-Forwarded-Host\":[\"10.0.10.XX:8000\"],\"X-Forwarded-Port\":[\"8000\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"28614ab9f2d1\"],\"X-Real-Ip\":[\"10.0.XX.XX\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"10.0.10.XX:8000\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"10.0.XX.XX:58506\",\"RequestURI\":\"/api/logs/stream?id=dfdf19c9d502\\u0026lastEventId=\",\"TLS\":null}"
time="2022-12-08T20:48:05Z" level=error msg="plugins-storage/sources/gop-1529375009/src/github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/bouncer.go:176:6: panic" plugin=plugin-bouncer module=github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin
time="2022-12-08T20:48:05Z" level=debug msg="Request has been aborted [10.0.XX.XX:58506 - /api/logs/stream?id=dfdf19c9d502&lastEventId=]: net/http: abort Handler" middlewareName=traefik-internal-recovery middlewareType=Recovery
time="2022-12-08T20:48:05Z" level=debug msg="vulcand/oxy/roundrobin/rr: begin ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/api/logs/stream\",\"RawPath\":\"\",\"OmitHost\":false,\"ForceQuery\":false,\"RawQuery\":\"id=28614ab9f2d1\\u0026lastEventId=\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"text/event-stream\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Accept-Language\":[\"fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Dnt\":[\"1\"],\"Pragma\":[\"no-cache\"],\"Referer\":[\"http://10.0.10.XX:8000/container/28614ab9f2d1\"],\"User-Agent\":[\"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36\"],\"X-Forwarded-Host\":[\"10.0.10.XX:8000\"],\"X-Forwarded-Port\":[\"8000\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"28614ab9f2d1\"],\"X-Real-Ip\":[\"10.0.XX.XX\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"10.0.10.XX:8000\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"10.0.XX.XX:58505\",\"RequestURI\":\"/api/logs/stream?id=28614ab9f2d1\\u0026lastEventId=\",\"TLS\":null}"
time="2022-12-08T20:48:05Z" level=debug msg="vulcand/oxy/roundrobin/rr: Forwarding this request to URL" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/api/logs/stream\",\"RawPath\":\"\",\"OmitHost\":false,\"ForceQuery\":false,\"RawQuery\":\"id=28614ab9f2d1\\u0026lastEventId=\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"text/event-stream\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Accept-Language\":[\"fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Dnt\":[\"1\"],\"Pragma\":[\"no-cache\"],\"Referer\":[\"http://10.0.10.XX:8000/container/28614ab9f2d1\"],\"User-Agent\":[\"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36\"],\"X-Forwarded-Host\":[\"10.0.10.XX:8000\"],\"X-Forwarded-Port\":[\"8000\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"28614ab9f2d1\"],\"X-Real-Ip\":[\"10.0.XX.XX\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"10.0.10.XX:8000\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"10.0.XX.XX:58505\",\"RequestURI\":\"/api/logs/stream?id=28614ab9f2d1\\u0026lastEventId=\",\"TLS\":null}" ForwardURL="http://172.18.0.2:8080"
```

> msg="Request has been aborted [10.0.XX.XX:58506 - /api/logs/stream?id=dfdf19c9d502&lastEventId=]: net/http: abort Handler" middlewareName=traefik-internal-recovery middlewareType=Recovery

**Comment by mpsarakis** [comment]
Hello Mathieu,

Glad you could reproduce the error!

So far and as far as I can see, errors in the log do not have any visible impacts on functionalities/navigation. They just appears in the logs.

Yes, maybe that's on Traefik side... strange combination with Dozzle...

Good luck to find the issue and do not hesitate if you need something from me.

Marios

**Comment by mathieuHa** [comment]
Hi Marios,

To follow Up I've found the following [issue](https://github.com/traefik/traefik/issues/8937) I believe to be connected in the Traefik repo.
I've added what we discovered and we will see if they find anything on their side.

Best,
Mathieu

**Comment by mpsarakis** [comment]
Hello Mathieu,

Great find, this looks definitely familiar to what we have experienced.

Hopefully they should be able to find a fix!

Thanks for the update.

Marios

**Comment by mathieuHa** [comment]
Hello @mpsarakis,

To keep you updated, they found a fix and it has been released in version v0.15.0 of Traefik/Yaegi [github.com/traefik/yaegi/releases/tag/v0.15.0](https://github.com/traefik/yaegi/releases/tag/v0.15.0)

It was integrated in Traefik 2.9 and 3.0 branch 3 weeks ago [github.com/traefik/traefik/commit/39331e41a85274fa24fbf282b0946462d5f2abbd](https://github.com/traefik/traefik/commit/39331e41a85274fa24fbf282b0946462d5f2abbd)

And in Traefik since 2.9.7! [github.com/traefik/traefik/releases/tag/v2.9.7](https://github.com/traefik/traefik/releases/tag/v2.9.7)

I'll close as completed now,
Again thanks for the feedback!

Best,
Mathieu

**Comment by mpsarakis** [comment]
Hello @mathieuHa,

Thanks for the update, nice it is fixed!

All the best.

Marios
